pkg/kube/build_lock.go

```diff
@@ -0,0 +1,415 @@
+package kube
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/jenkins-x/jx/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Labels required to be a lock. Anything else should be ignored.
+var buildLockLabels map[string]string = map[string]string{
+	"jenkins-x.io/kind": "build-lock",
+}
+
+// AcquireBuildLock acquires a build lock, to avoid other builds to edit the
+// same namespace while a deployment is already running, other deployments
+// can negotiate which one should run after, by editing its data.
+// Returns a function to release the lock (to be called in a defer)
+// Returns an error if a newer build is already running, or if an error happened
+func AcquireBuildLock(kubeClient kubernetes.Interface, devNamespace, namespace string) (func() error, error) {
+	// Only lock if running in Tekton
```

Comment:
this is a heck of a function, I am wondering whether it could not be split up a bit?

Reply:
Sure it can be, I'll work on that.